Migration guide from multi-cluster (#51316)
keycloak/keycloak: docs/guides/high-availability/multi-cluster-v2/migrate-from-v1-to-v2.adoc

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/links.adoc" as links>

<@tmpl.guide
title="Migrating from multi-cluster v1 to v2"
summary="Migrate an existing multi-cluster v1 deployment to v2 (stateless mode)."
tileVisible="false"
preview="true"
previewDiscussionLink="https://github.com/keycloak/keycloak/discussions/50640" >

This {section} describes how to migrate an existing <@links.ha id="multi-cluster-introduction" /> deployment to <@links.ha id="multi-cluster-v2-introduction" />.

Multi-cluster v2 uses the `stateless` feature to store volatile data in the database instead of an external {jdgserver_name} cluster.
This simplifies the deployment architecture by removing {jdgserver_name}, its cross-site replication, the fencing automation, and {jdgserver_name}-specific monitoring.

== Overview of changes

The migration involves the following changes:

.Components removed
[%autowidth]
|===
|Component |Reason

|External {jdgserver_name} cluster

> |Volatile data (authentication sessions, action tokens, login failure counters) is stored in the database instead.

Reviewer: Attempt to simplify it by not needing to explain or understand "Volatile data". And in stateless setups, it is no longer volatile.

```suggestion
|Authentication sessions, action tokens and login failure counters are stored in the database instead.
```

|No inter-site cache replication is needed; the database handles all cross-site data.

|{jdgserver_name} credentials and TLS secrets
|No longer required without the external {jdgserver_name} cluster.

|Fencing automation (for example, AWS Lambda)
|Split-brain detection and fencing for {jdgserver_name} are no longer necessary.

|{jdgserver_name}-specific monitoring and alerts
|Prometheus alerts for {jdgserver_name} cache health and cross-site connectivity are no longer needed.
|===

.Components retained
[%autowidth]
|===
|Component |Notes

|Database
|The existing synchronously replicated database is retained. Plan for approximately double the CPU usage and write IOPS compared to v1.

|Load balancer
|The external load balancer is retained.
Remove any health checks that target {jdgserver_name} endpoints.
The load balancer should probe the {project_name} `/lb-check` endpoint.
|===

.Configuration changes to {project_name}
[%autowidth]
|===
|Change |Details

|Enable the `stateless` feature
|Replaces the `multi-site` feature.

|Set a unique cluster name per site
|Required for cross-cluster cache invalidation via the database outbox.

|Increase the database connection pool size

> |Volatile data is now stored in the database, increasing load.

Reviewer: Avoiding "Volatile data" similar to the above

```suggestion
|Some data is now stored in database instead of Infinispan.
```

|===

== Prerequisites

* A running multi-cluster v1 deployment as described in the <@links.ha id="multi-cluster-introduction" /> {section}.
* Familiarity with the v2 architecture, its requirements, and limitations as described in the <@links.ha id="multi-cluster-v2-introduction" /> {section} and the <@links.ha id="multi-cluster-v2-concepts" /> {section}.
* A database sized for the increased load.
With v2, database CPU usage and write IOPS can roughly double.
Review and adjust the database instance size, storage IOPS, and connection pool sizing accordingly.
See the <@links.ha id="multi-cluster-concepts-database-connections" /> {section} for details.
* Network latency between database nodes that meets the requirements.
+
include::../partials/network-latency.adoc[]

== Impact on volatile data

Since user sessions are already stored in the database in multi-cluster v1, users remain logged in after migrating to v2.

The following volatile data is lost during the migration:

In-progress authentication sessions::
Users who are mid-login need to restart their login flow.

Login failure counters::
Brute-force detection counters are reset.

> Any ongoing lockouts are cleared.

Reviewer: Permanent lockouts are already stored in the database. Therefore refine this to temporary lockouts. 

```suggestion
Any ongoing temporary lockouts are cleared.
```

kubectl --namespace keycloak delete secret ispn-xsite-sa-token
kubectl --namespace keycloak delete secret xsite-keystore-secret
kubectl --namespace keycloak delete secret xsite-truststore-secret
----
+
Adjust the secret names to match your deployment.

.. Optionally, uninstall the {jdgserver_name} Operator if it is no longer needed for other purposes.
--

. Remove the fencing automation.
+
If using the AWS Lambda-based fencing as described in the <@links.ha id="multi-cluster-deploy-aws-accelerator-fencing-lambda" /> {section}:
+
--
.. Delete the AWS Lambda function.
.. Remove the Prometheus `AlertmanagerConfig` and `PrometheusRule` CRs that triggered the Lambda.
.. Remove the SNS topic and any related IAM roles.
--
+
Adjust these steps based on your fencing implementation.

. Remove {jdgserver_name}-specific monitoring.
+
Remove any Prometheus `ServiceMonitor` or `PodMonitor` resources that target the {jdgserver_name} deployment, as well as any Grafana dashboards for {jdgserver_name} cache metrics.

. Deploy the updated {project_name} CR to all sites:
+
[source,bash]
----
kubectl --namespace keycloak apply -f keycloak.yaml
----

> . Verify the deployment.

Reviewer: For the verification, could we point the the v2 guide instead? Or extract it to a partial? Note that my other PR might be updating the partial.

+
Confirm that the {project_name} deployment is ready:
+
[source,bash]
----
kubectl wait --for=condition=Ready keycloaks.k8s.keycloak.org/keycloak
kubectl wait --for=condition=RollingUpdate=False keycloaks.k8s.keycloak.org/keycloak
----
+
Afterward, verify that the external load balancer routes traffic to all sites.

> == Migrating on bare metal or virtual machines

Reviewer: We never had a guide for bare-metal multi-cluster v1. With that, can we skip the chapter "Migrating on bare metal or virtual machines"?
